### PR TITLE
Issue 835

### PR DESCRIPTION
**Pull request by squirrelo**
This fixes the analysis filepath issues and adds a test to make sure it doesn't happen again.

### CODE REVIEW

**Comment by ElDeveloper** [comment]
@squirrelo, there are legitimate flake8 errors in this build.

**Comment by squirrelo** [comment]
Not any more.

**Comment by ElDeveloper** [comment]
:boom: :boom:

File: qiita_db/test/test_analysis.py [hunk 1]
```diff
@@ -427,6 +428,26 @@ def test_build_files_raises_value_error(self):
         with self.assertRaises(ValueError):
             self.analysis.build_files(-10)
 
+    def test_add_file(self):
+        fp = join(get_mountpoint('analysis')[0][1], 'testfile.txt')
+        try:
```
Comment (ElDeveloper) [inline qiita_db/test/test_analysis.py]: If I understand correctly: if you can write to the file path you add it to the analysis and regardless of the result of these two lines you check if it exists and if it exists you remove it? Should that fp be removed in the teardown method of the test class? So we can guarantee that it is removed, the finally block here is kinda confusing.

**Comment by ElDeveloper** [comment]
This PR looks ok but I'm not super familiar with this codebase so it would be nice if someone else could review it in depth.

**Comment by squirrelo** [comment]
Yes, the finally makes it so the test always removes the file created
regardless of if the test errors or not. I can move it to teardown I guess,
I just hate having test specific things in teardown instead of in the test.

**Comment by ElDeveloper** [comment]
Ah got it. IMHO it's clearer if it's part of the teardown method (maybe
because that's what I'm used to see) than as an inline try/except block.
teardown methods are supposed to be dealing with these kind of things so
it's really up to you.

File: qiita_db/util.py [hunk 2]
```diff
@@ -608,8 +608,8 @@ def get_mountpoint(mount_type, conn_handler=None, retrieve_all=False):
             "SELECT data_directory_id, mountpoint, subdirectory FROM "
             "qiita.data_directory WHERE data_type='%s' and active=true"
             % mount_type)]
-
-    return [(d, join(get_db_files_base_dir(), m, s)) for d, m, s in result]
+    basedir = get_db_files_base_dir()
```
Comment (antgonza) [inline qiita_db/util.py]: I do not understand this change, is it required or is for clarity?

Comment (squirrelo) [inline qiita_db/util.py]: The function does an SQL select every time it's called, so this makes it happen 1 time instead of N times.

Comment (antgonza) [inline qiita_db/util.py]: Excellent point! Thanks for catching that.

Comment (josenavas) [inline qiita_db/util.py]: Should this use `get_db_files_base_dir` or should we move to the `get_mountpoint` function that @antgonza added a while ago?

Comment (antgonza) [inline qiita_db/util.py]: This line is within the get_mountpoint function ....

File: qiita_db/analysis.py [hunk 3]
```diff
@@ -753,7 +753,7 @@ def _add_file(self, filename, filetype, data_type=None, conn_handler=None):
                "checksum, checksum_algorithm_id, data_directory_id) VALUES "
                "(%s, %s, %s, %s, %s) RETURNING filepath_id")
         # magic number 1 is for crc32 checksum algorithm
-        fpid = conn_handler.execute_fetchone(sql, (fullpath, fptypeid,
+        fpid = conn_handler.execute_fetchone(sql, (filename, fptypeid,
```
Comment (josenavas) [inline qiita_db/analysis.py]: This line looks ok. However, I'm looking at the function in general `_add_file` and it looks pretty similar to `insert_filepaths` in the `util.py` file. Is it possible to remove this one and use the `insert_filepaths` function here, so no code duplication happens in any place? BTW, I would say that the code duplication was the source of this bug...

File: qiita_db/support_files/patches/14.sql [hunk 4]
```diff
@@ -0,0 +1,2 @@
+--Placeholder for python to run
```
Comment (josenavas) [inline qiita_db/support_files/patches/14.sql]: Can you add the date to the patch?

File: qiita_db/support_files/patches/python_patches/14.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+from qiita_db.sql_connection import SQLConnectionHandler
+from os.path import basename
+
+# This changes all analysis files to be relative path instead of absolute
```
Comment (josenavas) [inline qiita_db/support_files/patches/python_patches/14.py]: Add date and move all comments to the first line of the file (for consistency with other patches)

File: qiita_db/util.py (hunk 2 above)
Comment (josenavas) [inline qiita_db/util.py]: Ooops, I just saw that this is actually the `get_mountpoint function`... :sweat_smile::innocent:

**Comment by josenavas** [comment]
Few comments @squirrelo

File: qiita_db/support_files/patches/python_patches/14.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+conn_handler = SQLConnectionHandler()
+
+filepaths = conn_handler.execute_fetchall(
+    'SELECT f.* from qiita.filepath f JOIN qiita.analysis_filepath afp ON '
+    'f.filepath_id = afp.filepath_id')
+for filepath in filepaths:
+    filename = basename(filepath['filepath'])
```
Comment (wasade) [inline qiita_db/support_files/patches/python_patches/14.py]: Just to verify: these files are still relative to a unique directory associated with a study or a meta-analysis, correct?

File: qiita_db/analysis.py (hunk 3 above)
Comment (squirrelo) [inline qiita_db/analysis.py]: I can for the filepath insertion, but I still need it for the linking of analysis to the filepath.

Comment (josenavas) [inline qiita_db/analysis.py]: Ok, if you can use `insert_filepaths` to do part of the work here, use it, so we avoid code duplication. If you still need some code to finish other work it is fine to do it here.

File: qiita_db/support_files/patches/14.sql (hunk 4 above)
Comment (squirrelo) [inline qiita_db/support_files/patches/14.sql]: done

**Comment by antgonza** [comment]
Pull from upstream master.

**Comment by squirrelo** [comment]
ready for review again.

**Comment by antgonza** [comment]
The patch is not working in my machine because the mountpoint has an ending / and the file path doesn't
```
Traceback (most recent call last):
  File "qiita_db/support_files/patches/python_patches/14.py", line 23, in <module>
    mp_id = mountpoints[dirname(filepath['filepath'])]
KeyError: '/Users/antoniog/Desktop/CurrentWork/largeFiles/qiita/base_dir/analysis'
```
and I have these values in `filepath`, `mountpoints`
```
[308L, '/Users/antoniog/Desktop/CurrentWork/largeFiles/qiita/base_dir/analysis/1_analysis_mapping.txt', 9L, '833932644', 1L, 1L]
{'/Users/antoniog/Desktop/CurrentWork/largeFiles/qiita/base_dir/analysis/': 1L}
```

**Comment by wasade** [comment]
Should the mount points have a trailing '/'? That technically refers to the _contents_ of the path, not that path

**Comment by antgonza** [comment]
Got it but the current configuration allows you to have either so
validating both cases or not allowing one in the configuration should
be the way to go.

**Comment by antgonza** [comment]
:+1:

**Comment by antgonza** [comment]
@squirrelo conflicts, @ElDeveloper could you merge after conflicts are solved? Thanks.

File: qiita_db/support_files/patches/14.sql (hunk 4 above)
Comment (ElDeveloper) [inline qiita_db/support_files/patches/14.sql]: Haha, either git or the conflict resolution got messed, this change should have only gone in 15.sql ... I'll let it slide :laughing:

**Pull request merged**